src/passes/LLVMNontrappingFPToIntLowering.cpp

```diff
@@ -0,0 +1,185 @@
+// is no trap, but need not ensure any particular result. The transformation
+// in this pass is the same as the one used by LLVM to lower fptoui/fptosi
+// to wasm trapping instructions.
+
+// For example, if a conversion is guarded by a range check in the source, LLVM
+// can move the conversion before the check (and instead guard the use of the
+// result, which may be poison). This is valid in LLVM and for the nontrapping
+// wasm fptoint instructions but not for the trapping conversions. The
+// transformation in this pass is valid only if the nontrapping conversions
+// in the wasm were generated from LLVM and implement LLVM's conversion
+// semantics.
+
+namespace wasm {
+struct LLVMNonTrappingFPToIntLoweringImpl
+  : public WalkerPass<PostWalker<LLVMNonTrappingFPToIntLoweringImpl>> {
+  bool isFunctionParallel() override { return true; }
+
+  std::unique_ptr<Pass> create() override {
+    return std::make_unique<LLVMNonTrappingFPToIntLoweringImpl>();
+  }
+
+  UnaryOp getReplacementOp(UnaryOp op) {
+    switch (op) {
+      case TruncSatSFloat32ToInt32:
+        return UnaryOp::TruncSFloat32ToInt32;
+      case TruncSatUFloat32ToInt32:
+        return UnaryOp::TruncUFloat32ToInt32;
+      case TruncSatSFloat64ToInt32:
+        return UnaryOp::TruncSFloat64ToInt32;
+      case TruncSatUFloat64ToInt32:
+        return UnaryOp::TruncUFloat64ToInt32;
+      case TruncSatSFloat32ToInt64:
+        return UnaryOp::TruncSFloat32ToInt64;
+      case TruncSatUFloat32ToInt64:
+        return UnaryOp::TruncUFloat32ToInt64;
+      case TruncSatSFloat64ToInt64:
+        return UnaryOp::TruncSFloat64ToInt64;
+      case TruncSatUFloat64ToInt64:
+        return UnaryOp::TruncUFloat64ToInt64;
+      default:
+        WASM_UNREACHABLE("Unexpected opcode");
+    }
+  }
+
+  template<typename From, typename To> void replaceSigned(Unary* curr) {
+    BinaryOp ltOp;
+    UnaryOp absOp;
+    switch (curr->op) {
+      case TruncSatSFloat32ToInt32:
+      case TruncSatSFloat32ToInt64:
+        ltOp = BinaryOp::LtFloat32;
+        absOp = UnaryOp::AbsFloat32;
+        break;
+      case TruncSatSFloat64ToInt32:
+      case TruncSatSFloat64ToInt64:
+        ltOp = BinaryOp::LtFloat64;
+        absOp = UnaryOp::AbsFloat64;
+        break;
+      default:
+        WASM_UNREACHABLE("Unexpected opcode");
+    }
+
+    Builder builder(*getModule());
+    Index v = Builder::addVar(getFunction(), curr->value->type);
+    // if fabs(operand) < INT_MAX then use the trapping operation, else return
+    // INT_MIN
+    replaceCurrent(builder.makeIf(
+      builder.makeBinary(
+        ltOp,
+        builder.makeUnary(
+          absOp, builder.makeLocalTee(v, curr->value, curr->value->type)),
+        builder.makeConst(static_cast<From>(std::numeric_limits<To>::max()))),
+      builder.makeUnary(getReplacementOp(curr->op),
+                        builder.makeLocalGet(v, curr->value->type)),
+      builder.makeConst(std::numeric_limits<To>::min())));
```

Comment:
Are this value and '0' in `replaceUnsigned` arbitrarily chosen?
(LLVM's choices are different (but it looks they are arbitrarily chosen too): https://github.com/llvm/llvm-project/blob/2906fcadb8563a02949f852867cebc63e0539cb7/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp#L492)

And these values don't seem to conform to the real "saturating" spec, because we don't check whether the value overflows or underflows and just use a single substitute value. Is this the reason why this pass is LLVM-specific? Then what is the reason for not making pass a generic conversion pass by implementing the real saturation behavior? (I guess performance?)

Reply:
For some explanation on this, see the comment at the top of this file. The short answer is that this is UB under LLVM's semantics so it doesn't matter. And yes, this is why the pass is LLVM-specific. The reason is basically performance (a faithful implementation would have more checks than is necessary for LLVM) and simplicity of the transform. The other reason is that if there happens to be code in the world that depends on LLVM's current behavior (UB notwithstanding), then it makes sense to have this transform just take us back to LLVM's current behavior. 
And, regarding your other comment above: Yes, you are right that this doesn't handle the case of INT_MIN+1. I didn't notice that when I wrote this, but did notice it when we were fixing the UB in the SSE tests. I decided that it's probably better just to preserve LLVM's existing behavior here than to change it. Most users will now just be getting the default more-correct saturating behavior anyway.